Support configuring modules through imports

jathak opened the pull request:
Fixes #882

See sass/sass-spec#1497

jathak commented:
I think that makes sense, but could there ever be a case where the `Environment` is mutated after the `Configuration` is created but before it's used?

nex3 commented:
Good question... I guess that does run into the problem where variables defined in the midstream file suddenly become part of the upstream file's configuration, which we don't want, so copying everything over is probably better in the end.

nex3 commented on lib/src/async_environment.dart:
```diff
@@ -670,6 +671,22 @@ class AsyncEnvironment {
+          _variableNodes == null ? <String, AstNode>{} : _variableNodes[i];
+      for (var name in values.keys) {
+        configuration[name] =
+            ConfiguredValue(values[name], nodes[name]?.span, nodes[name]);
```
Materializing `AstNode`s' spans can be expensive, which is why we pass them around as `AstNode`s in the first place, so I'd like to avoid `nodes[name]?.span` here. In practice this span is also never used, since an implicit configuration's variables are never invalid.

We could just pass `null` here, or we could encode this more statically by having a subclass with a constructor `ImplicitConfiguredValue(Value, AstNode)` and defining `Configuration.implicit(Map<String, ImplicitConfiguredValue>)`. Up to you.

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+  /// A map from variable names (without `$`) to values.
+  ///
+  /// When this is empty, it may be unmodifiable, so [Map.remove] should not be
+  /// called on this.
```
Now that we have a full `Configuration` class, we can enforce this more aggressively by only exposing an unmodifiable view of this map and exposing a `Configuration.remove()` method that handles empty configurations correctly.

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+  final Map<String, ConfiguredValue> values;
+
+  /// Whether or not this configuration is implicit.
+  final bool isImplicit;
```
I might move the discussion of what an implicit configuration means from the class documentation to this getter.

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+        isImplicit = false;
+
+  bool get isEmpty => values.isEmpty;
+  bool get isNotEmpty => values.isNotEmpty;
```
Typically classes that aren't implementing core library interfaces don't define `isNotEmpty`. (Defining `isNotEmpty` at all was a design mistake—having a different way to negate exactly one boolean getter doesn't make sense—but there's no going back on that for the core libraries at this point.)

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+  bool get isNotEmpty => values.isNotEmpty;
+
+  /// Creates a new configuration from this one based on a `@forward` rule.
+  Configuration throughForward(ForwardRule forward) {
```
Moving this logic here is nice.

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+
+  /// Creates a copy of this configuration.
+  Configuration clone() => isEmpty
+      ? const Configuration.empty()
```
This will return an explicit configuration even if `this` is implicit. Does that matter? Maybe all empty configurations should be considered implicit, since they don't cause duplicate-configuration errors?

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+}
+
+/// A variable value that's been configured using `@use ... with`.
+class ConfiguredValue {
```
I usually put every class definition in its own file, even if two classes are related, unless one is private or they have to access private members from one another.

nex3 commented on lib/src/configuration.dart:
```diff
@@ -0,0 +1,85 @@
+      : Configuration({...values}, isImplicit: isImplicit);
+}
+
+/// A variable value that's been configured using `@use ... with`.
```
Some values are configured without `@use ... with`—maybe just say "configured for a `Configuration`"?

nex3 commented on lib/src/visitor/async_evaluate.dart:
```diff
@@ -576,22 +574,21 @@ class _EvaluateVisitor
     var alreadyLoaded = _modules[url];
     if (alreadyLoaded != null) {
-      if ((configuration ?? _configuration).isNotEmpty) {
+      configuration ??= _configuration;
```
I don't think you want to make this assignment. We use the fact that this was passed as null later on to determine whether to throw errors about unused configuration variables.

nex3 commented on lib/src/visitor/async_evaluate.dart:
```diff
@@ -635,8 +632,7 @@ class _EvaluateVisitor
       if (configuration != null) {
-        _configuration =
-            configuration.isEmpty ? const {} : Map.of(configuration);
+        _configuration = configuration.clone();
```
Style nit: you can make this a single-line `if` now.

nex3 commented on lib/src/visitor/async_evaluate.dart:
```diff
@@ -661,11 +657,11 @@ class _EvaluateVisitor
       if (configuration != null && _configuration.isNotEmpty) {
```
Shouldn't this also check whether the configuration is implicit?

nex3 commented on lib/src/visitor/async_evaluate.dart:
```diff
@@ -1306,13 +1285,20 @@ class _EvaluateVisitor
 
+        // This configuration is only used if it passes through a `@forward`
+        // rule, so we avoid creating unnecessary ones for performance reasons.
+        if (stylesheet.forwards.isNotEmpty) {
```
Before you submit, try removing this condition and *always* creating the environment, and run the result against sass-spec. It shouldn't change the behavior, so if it does you'll have found a bug.

jathak commented:
Okay, I think I've addressed all comments. I also tested with the condition for creating the implicit configuration removed and confirmed that all of the specs still pass.
